src/common/exception.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <cpptrace/cpptrace.hpp>
+#include <exception>
+#include <string>
+#include <system_error>
+
+// Custom exception implementations that store the stacktrace of the exception
+// at the call-site that it was thrown at. This implementation is based off
+// cpptrace's implementation and we also use cpptrace as a library for
+// generating cross-platform stack traces.
+//
+// Our implementatin is the same as cpptrace with the added feature that we dump
+// the stack trace with inline code-snippets for the call-site that threw the
+// exception instead of just the stack trace itself.
+
+namespace oxen {
+
+// Catches application termination and dumps a stack-trace where possible.
+// It must be registered as the handler on startup, preferrably in main() e.g.
+//
+//   std::set_terminate(oxen::on_terminate_handler)
+//
+// This implementation is mostly copied from cpptrace's
+// `register_terminate_handler` except that we look for `oxen::` exceptions
+// thrown from one of the types in this file.
+void on_terminate_handler();
+
+struct exception : public std::exception {
+    exception(std::string&& msg,
+            cpptrace::raw_trace&& trace =
+                    cpptrace::generate_raw_trace(/*skip*/ 0, /*max_depth*/ 256));
+
+    const char* what() const noexcept override;
+
+    const cpptrace::raw_trace& trace() const noexcept;
+
+    cpptrace::raw_trace raw_trace;
+
+    // Cache of the string to be printed when 'what' is called because we store
+    // the stack trace as a list of U64 addresses.
+    mutable std::string what_msg;
+
+    // The message created when the exception was thrown.
+    std::string user_msg;
+};
```

Comment:
This feels fairly heavy -- I think it should be turned on only for a debug build, or possibly behind a separate cmake option that we turn off for release builds.

Secondly, this (nor the cpptrace version of the exception types) doesn't preserve the std::exception hierarchy, which has the potential to cause bugs in unexpected ways.  For example, code such as:
```C++
    try { ... }
    catch (const std::invalid_argument& e) { ... }
    catch (const std::exception& e) { ... }
```
would be broken by throwing a oxen::invalid_argument because the expected C++ exception hierarchy

    std::invalid_argument -> std::logic_error -> std::exception

doesn't get preserved here, but instead we have:

    oxen::invalid_argument -> oxen::exception -> std::exception

This is fixable, though, with something like this:
```C++
    // oxen::exception extends a standard exception object adding stack trace information to it, when
    // in a debug build.  For release builds, the oxen::exception<T> is simply a typedef for a T.

    #ifdef NDEBUG

    template <std::derived_from<std::exception> StdExcept>
    using exception = StdExcept;

    #else

    template <std::derived_from<std::exception> StdExcept>
    class exception : public StdExcept {
      public:
        template <typename... Args>
            requires std::constructible_from<StdExcept, Args...>
        exception(Args&&... args,
            cpptrace::raw_trace&& trace = cpptrace::generate_raw_trace(/*skip*/ 0, /*max_depth*/ 256))
        : StdException{std::forward<Args>(args)...}, raw_trace{std::move(trace)} {}

        const char* what() const noexcept override {
            if (what_msg.empty())
                what_msg = make_traced_msg(StdExcept::what(), raw_trace);
            return what_msg.c_str();
        }

        const cpptrace::raw_trace& trace() const noexcept { return raw_trace; }

      private:
        // Cache of the string to be printed when 'what' is called because we store
        // the stack trace as a list of U64 addresses and only construct the full
        // trace string on demand.
        mutable std::string what_msg;

        cpptrace::raw_trace raw_trace;
    };

    std::string make_traced_msg(const char* exc_what, const cpptrace::raw_trace& trace) {
        // ... construct combined exception message
    }
```
and drop all of the oxen::logic_error, oxen::domain_error and so on; anywhere you want to throw would now be:
```c++
    throw oxen::exception<std::runtime_error>{"whatever"};
```
I'm even tempted to suggest adding a global namespace typedef for this such as:
```C++
    template <std::derived_from<std::exception> StdExcept>
    using traced = oxen::exception<StdExcept>;
```
so that we can just write `throw traced<std::runtime_error>{"whatever"};` everywhere, which has a nice, easy-to-remember look to it.